src/GraphQL/Value.hs

```diff
@@ -79,24 +86,43 @@ instance ToJSON List where
 --
 -- TODO: the "map" needs to keep track of insertion order so we can
 -- return fields in query order as mandated by the spec.
-newtype Map = Map (Map.Map Name GraphQL.Value.Value) deriving (Eq, Ord, Show)
+newtype Map = Map [(Name,  GraphQL.Value.Value)] deriving (Eq, Ord, Show, Generic, Monoid)
+
+
+mapFromList :: [(Name,  GraphQL.Value.Value)] -> Map
+mapFromList = Map
+
+-- TODO this wouldbe nicer with a prism `_ValueMap` but don't want to
+-- pull in lens as dependency.
+unionMap :: [Value] -> Either Text Value
+unionMap values = map (ValueMap . fold)  (sequence (map isValueMap values))
+  where
+    isValueMap = \case
+      (ValueMap m) -> Right m
+      _ -> Left "non-ValueMap member"
+
 
 instance ToJSON Map where
-  toJSON (Map x) = toJSON x
+  -- Direct encoding to preserve order of keys / values
+  toJSON (Map xs) = toJSON (Map.fromList xs)
+  toEncoding (Map xs) = pairs (fold (map (\(k, v) -> (toS k) .= v) xs))
 
 
 data Field = Field Name GraphQL.Value.Value deriving (Eq, Show, Ord)
 
 makeField :: (StringConv name Name, ToValue value) => name -> value -> Field
 makeField name value = Field (toS name) (toValue value)
 
+-- TODO I'm unclear about what we're going to use FieldSet for. Do we
+-- need it?
```

Comment:
I don't _think_ so. I think it was meant to be a realized selection set, but I can't be sure. Delete it, and if we discover we need it, we can add something similar & better later.